src/circuit/gadgets/constraints.rs

```diff
@@ -430,6 +503,444 @@ pub fn must_be_simple_bit(x: &Boolean) -> AllocatedBit {
+        &mut cs.namespace(|| "b is not zero and cond"),
+        &b_is_zero.not(),
+        &cond,
+    )?;
+    enforce_implication(
+        &mut cs.namespace(|| "enforce u64 mod decomposition"),
+        &b_is_not_zero_and_cond,
+        &u64mod_decomp,
+    )?;
+
+    enforce_less_than_bound(
+        &mut cs.namespace(|| "remainder in range b"),
+        cond,
+        alloc_r.clone(),
+        alloc_arg2.clone(),
+    )?;
+
+    let arg1_u64_tag = alloc_equal(
+        &mut cs.namespace(|| "arg1 u64 tag"),
+        alloc_arg1.tag(),
+        &g.u64_tag,
+    )?;
+    let arg2_u64_tag = alloc_equal(
+        &mut cs.namespace(|| "arg2 u64 tag"),
+        alloc_arg2.tag(),
+        &g.u64_tag,
+    )?;
+    enforce_true(&mut cs.namespace(|| "check arg1 u64 tag"), &arg1_u64_tag)?;
+    enforce_true(&mut cs.namespace(|| "check arg2 u64 tag"), &arg2_u64_tag)?;
+
+    Ok((alloc_q, alloc_r))
+}
+
+// Enforce the num < b. This is done by proving (b - num) is positive.
+pub fn enforce_less_than_bound<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    cond: Boolean,
+    num: AllocatedPtr<F>,
+    bound: AllocatedPtr<F>,
+) -> Result<(), SynthesisError> {
+    let diff_bound_num = sub(
+        &mut cs.namespace(|| "bound minus num"),
+        bound.hash(),
+        num.hash(),
+    )?;
+    let double_diff_bound_num = add(
+        &mut cs.namespace(|| "double diff bound num"),
+        &diff_bound_num,
+        &diff_bound_num,
+    )?;
+    let double_diff_bound_num_bits = double_diff_bound_num
+        .to_bits_le(&mut cs.namespace(|| "diff bits"))
+        .unwrap();
+    let lsb_2diff_bound_num = double_diff_bound_num_bits.get(0);
+    let diff_bound_num_is_negative = lsb_2diff_bound_num.unwrap();
+
+    enforce_implication(
+        &mut cs.namespace(|| "enforce u64 range"),
+        &cond,
+        &diff_bound_num_is_negative.not(),
+    )
+}
+
+// Convert from bn to num. This allocation is NOT constrained here.
+// In the circuit we use it to prove u64 decomposition, since using bn
+// we have division with remainder, which is used to find the quotient
+// after dividing by 2ˆ64. Therefore we constrain this relation afterwards.
+pub fn bn_to_num<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    bn: BigUint,
+) -> Result<AllocatedNum<F>, SynthesisError> {
+    let bytes_le = bn.to_bytes_le();
+    let mut bytes_padded = [0u8; 32];
+    bytes_padded[0..bytes_le.len()].copy_from_slice(&bytes_le);
+    let ff = F::from_bytes(&bytes_padded).unwrap();
+    let num = AllocatedNum::alloc(&mut cs.namespace(|| "num"), || Ok(ff)).unwrap();
+    Ok(num)
+}
+
+// Convert from num to u64.
+pub fn u64_op<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    g: &GlobalAllocations<F>,
+    maybe_u64: &AllocatedPtr<F>,
+    store: &Store<F>,
+) -> Result<AllocatedPtr<F>, SynthesisError> {
+    let p64_bn = match g.power2_64_num.get_value() {
+        Some(v) => BigUint::from_bytes_le(v.to_repr().as_ref()),
```

Comment:
This seems very roundabout. Let's not desensitize ourselves to all the weird thing being done here if we could just use a `BigUint` constant. Probably we should define such a constant, calculated explicitly, then also use that to populate the global. Then I think we should be able to just use the constant here.

Reply:
removed, since the constant was already there in the globals